### DISCUSSION TITLE
Choosing a device for Direct ML provider

### DISCUSSION THREAD

**Discussion by smirnov-alexey** [microsoft/onnxruntime#14840] [Other Q&A]
Hi! I'm writing a C++ sample using ONNXRT only and trying to utilize Direct ML execution provider. I have a couple of discrete GPUs and an integrated one. How do I specify which device I want to use for ONNXRT and/or DML? How to discover devices except for `get_device()`? Thank you!

**Comment by deischi** [comment]
To specify which device to use, you can use `SessionOptionsAppendExecutionProvider_DML`. 
This is the code I am using to set the appropiate sessionsOptions:
```
const OrtApi& ortApi = Ort::GetApi();
const OrtDmlApi* ortDmlApi = nullptr;
Ort::ThrowOnError(ortApi.GetExecutionProviderApi("DML", ORT_API_VERSION, reinterpret_cast<const void**>(&ortDmlApi)));
Ort::ThrowOnError(ortDmlApi->SessionOptionsAppendExecutionProvider_DML(sessionOptions, iAdapterNr));

```

To discover which devices are possible, you can use DXGIs `EnumerateAdapters1`.
Here is a some excerpt of the code I use to iterate over the available adapters - this example will search for the nVidia adapter:
```
CComPtr<IDXGIFactory2> pDXGIFactory;
::CreateDXGIFactory1(__uuidof(IDXGIFactory2), (void**)&pDXGIFactory);

int iAdapterNr = 0;
int iBestAdapter  = 0;
for (CComPtr<IDXGIAdapter1> pAdapter; pDXGIFactory->EnumAdapters1(iAdapterNr, &pAdapter) != DXGI_ERROR_NOT_FOUND; iAdapterNr++, pAdapter = nullptr)
{
	DXGI_ADAPTER_DESC1 desc = { 0 };
	pAdapter->GetDesc1(&desc);

	if(desc.VendorId == 0x10de)
	{
		iBestAdapter = iAdapterNr;
	}
}
return iBestAdapater;
```
